core/src/main/scala/spinal/core/Enum.scala

```diff
@@ -349,6 +349,29 @@ object binaryOneHot extends SpinalEnumEncoding{
   setName("oh")
 }
 
+/**
+ * Gray encoding (sequentially assigned)
+ * @example{{{ 000, 001, 011, 010, ... }}}
+ * @note If used in FSM is is not ensured that only gray encoding preserving
+ *       transitions are done. If that is needed e.g. for CDC reasons, the
+ *       transitions must be checked manually.
+ */
+object graySequential extends SpinalEnumEncoding {
+  private def gray2binary(gray: BigInt) = {
+    assert(gray.bitLength <= 16, "conversion currently only implemented for <= 16 bit")
```

Comment:
Maybe it's better to keep 16 a variable but constant.

Reply:
Or define a generic function?

```scala

object Test extends App {
  private def gray2binary(gray: BigInt) = {
    @tailrec def aux(g: BigInt, b: BigInt): BigInt = {
      if (g > 0) {
        val g1 = g >> 1
        aux(g1, b ^ g1)
      } else {
        b
      }
    }
    aux(gray, gray)
  }
  private def binary2gray(bin: BigInt) = bin ^ (bin >> 1)

  // helpers for tests
  def assertions(i: BigInt): Unit = {
    assert(gray2binary(binary2gray(i)) == i)
    assert(binary2gray(gray2binary(i)) == i)

    if (i > 0)
      assert(isPow2(binary2gray(i) ^ binary2gray(i - 1)))
  }
  // basic test
  for (n <- 0 until 1000) {
    val i = BigInt(n)

    def toString(x: BigInt): String = {
      val digits = x.toString(2)
      val n = 4
      val l = digits.length
      if (l <= n) {
        "0" * (n - l) + digits
      } else digits
    }
    if (i < 16)
      println((toString(i), toString(binary2gray(i))))

    assertions(i)
  }

  // random test
  for (n <- 0 until 1000){
    def r: BigInt = {
      val x = scala.util.Random.nextLong()
      if (x < 0) r
      else BigInt(x)
    }
    assertions(r)
  }
}
```

Reply:
Also, maybe these two helper functions should be available, in `misc` for instance?

Reply:
are we really discussing about the finer detail of having more than 2^16 enum values / fsm states? xD

Reply:
but no worries, easy to extend